components/brave_wallet/browser/eth_allowance_manager.cc

```diff
@@ -0,0 +1,403 @@
+#include "brave/components/brave_wallet/common/hash_utils.h"
+#include "brave/components/brave_wallet/common/hex_utils.h"
+#include "components/prefs/pref_service.h"
+#include "components/prefs/scoped_user_pref_update.h"
+#include "ui/base/l10n/l10n_util.h"
+
+namespace {
+constexpr char kLastBlockNumber[] = "last_block_number";
+constexpr char kAllowanceFound[] = "allowances_found";
+constexpr char kApproverAddress[] = "approver_address";
+constexpr char kContractAddress[] = "contract_address";
+constexpr char kSpenderAddress[] = "spender_address";
+constexpr char kAmount[] = "amount";
+
+constexpr char kAddress[] = "address";
+constexpr char kTopics[] = "topics";
+constexpr char kFromBlock[] = "fromBlock";
+constexpr char kToBlock[] = "toBlock";
+constexpr char kEarliestBlock[] = "earliest";
+constexpr char kLatestBlock[] = "latest";
+
+constexpr char kApprovalTopicHash[] = "Approval(address,address,uint256)";
+
+// Returns vector of chain_ids supported by allowance discovering.
+const std::vector<std::string>& GetChainIdsForAlowanceDiscovering() {
+  static base::NoDestructor<std::vector<std::string>> supported_chain_ids(
+      {brave_wallet::mojom::kMainnetChainId,
+       brave_wallet::mojom::kPolygonMainnetChainId,
+       brave_wallet::mojom::kAvalancheMainnetChainId,
+       brave_wallet::mojom::kCeloMainnetChainId,
+       brave_wallet::mojom::kArbitrumMainnetChainId,
+       brave_wallet::mojom::kOptimismMainnetChainId});
+  return *supported_chain_ids;
+}
+static std::string GetAllowanceMapKey(const std::string& contract_address,
+                                      const std::string& approver_addr,
+                                      const std::string& spender_address) {
+  return base::JoinString({contract_address, approver_addr, spender_address},
+                          "_");
+}
+}  // namespace
+
+namespace brave_wallet {
+
+EthAllowanceManager::EthAllowanceManager(BraveWalletService* wallet_service,
+                                         JsonRpcService* json_rpc_service,
+                                         KeyringService* keyring_service,
+                                         PrefService* prefs)
+    : wallet_service_(wallet_service),
+      json_rpc_service_(json_rpc_service),
+      keyring_service_(keyring_service),
+      prefs_(prefs),
+      weak_ptr_factory_(this) {}
+
+EthAllowanceManager::~EthAllowanceManager() = default;
+
+/**
+ * @brief Represents start of the allowance discovering operation
+ *
+ * @param callback callback to front end, to pass allowances list, error code
+ * and description
+ */
+void EthAllowanceManager::DiscoverEthAllowancesOnAllSupportedChains(
+    BraveWalletService::DiscoverEthAllowancesCallback callback) {
+  if (is_eth_allowance_discovering_running_) {
+    std::move(callback).Run(
+        {}, mojom::ProviderError::kInternalError,
+        l10n_util::GetStringUTF8(IDS_WALLET_ALREADY_IN_PROGRESS_ERROR));
+    return;
+  }
```

Comment:
Let's keep list of callbacks. When result is ready - push it to all stored callbacks.